### ISSUE TITLE
Feature-request: Support for rerunning failed flaky tests

### ISSUE THREAD

**Issue by arcuri82** [junit-team/junit5#1558]
## Overview

Feature request to add native support to rerun failing flaky tests

**Feature request** 

Is there any plan to support the ability of rerunning failed flaky tests? For example, it would be great to have something like `@Flaky(rerun=3)`, which would rerun a failing flaky test up to n times.

In the *past*, in Surefire/Failsafe we could use `rerunFailingTestsCount`. However, it does not work in JUnit 5. As JUnit 5.0.0 was released nearly 1 year ago, it is unclear if and when it will be supported in Maven.

And, even if it will be supported in Maven in the future, it could be good to have a special annotation to specify that only some tests are expected to be flaky, and not *all* of them.

At the moment, the workaround is to stick with JUnit 4, which is a kind of a shame as JUnit 5  has a lot of interesting features :(  (which I can only use in projects with no flaky tests)

Related [SO question](https://stackoverflow.com/questions/46181026/junit5-how-to-repeat-failed-test).

**Comment by sbrannen** [comment]
No, to my knowledge this are not currently any plans to implement such a feature.

But... I think it's a very good idea.

I've often wanted something like [Spring Retry's `@Retryable`](https://github.com/spring-projects/spring-retry#java-configuration-for-retry-proxies) support for test methods. Of course, Spring Retry supports a much larger feature set than we would ever need/want in JUnit Jupiter.

Thanks for providing the link to the discussion on Stack Overflow. There are two interesting implementations there that can serve as inspiration.

Ideally, I'd like to see something like a `@RetryableTest` annotation that allows one to specify:

- maximum number of retries
- configurable exceptions that initiate a retry, defaulting to `{Throwable.class}`
- potentially a minimum success rate

@junit-team/junit-lambda, thoughts?

**Comment by sbrannen** [comment]
Tentatively slated for 5.4 M1 for _team discussion_

**Comment by jbduncan** [comment]
> configurable exceptions that initiate a retry, defaulting to `{Throwable.class}`

Hmm, how about defaulting to `Exception.class` or `Throwable.class-except-blacklisted-exceptions`?

My thinking is that in the off-chance an `OutOfMemoryError` is thrown, it should really be propagated to allow the JVM to shutdown as cleanly as it possibly can. :thinking:

**Comment by sormuras** [comment]
I don't like flaky tests (read: checks). Who does?
I don't like the idea of _"fixing"_ flaky tests by (naiv, smart, conditional, [what|for]-ever) re-execution.
I don't want to support that in Jupiter.

Easy work-around:
```
@Test void test() { REPEAT flaky() UNTIL flaky-signals-green OR break-out }
boolean flaky() { ... return true : false ... }
```

Perhaps (!) such an annotation could live in an extension providing library: like https://junit-pioneer.org or another one. If we need to expose a new extension point tailored to make the "retry" functionality possible, well, I'm fine with that.

**Comment by arcuri82** [comment]
@sormuras I do not like flaky tests either :) but unfortunately they are a reality :(

not an issue with unit tests (unless you work with randomized algorithms), but they are a big problem for E2E tests. There, it is hard to escape from flakyness. Even if a test wrongly fails 1 out 1000 times, when you have hundreds/thousands of E2E tests, there is a high probability that at least 1 fails, failing the whole build.

Of course one can use a try/catch in a loop, but that is a lot of boilerplate that has to be repeated each time for each test.

In the past, it was not an issue with JUnit 4, as this was externally handled with Maven. But this is not the case any more. In my case, this is a major showstopper for using JUnit 5. In some projects we tried JUnit 5 and were forced to revert to 4. In others, we use JUnit 5 for unit tests, but then have a (Maven) module for the E2E using JUnit 4.

**Comment by nicolaiparlog** [comment]
Coincidentally, I recently hacked [an extension that repeats failed tests](https://github.com/junit-pioneer/junit-pioneer/commit/15ab0a6c97254335d877aa5587fc519ed8abf7f3#diff-94a60a248ed1fcd2fa7028a25ea6664e):

```java
class RepeatFailedTestTests {

	private static int FAILS_ONLY_ON_FIRST_INVOCATION;

	@RepeatFailedTest(3)
	void failsNever_executedOnce_passes() { }

	@RepeatFailedTest(3)
	void failsOnlyOnFirstInvocation_executedTwice_passes() {
		FAILS_ONLY_ON_FIRST_INVOCATION++;
		if (FAILS_ONLY_ON_FIRST_INVOCATION == 1) {
			throw new IllegalArgumentException();
		}
	}

	@RepeatFailedTest(3)
	void failsAlways_executedThreeTimes_fails() {
		throw new IllegalArgumentException();
	}

}
```

If JUnit 5 decides that flaky tests have no place in a proper testing framework, we don't have such scruples. :wink:

**Comment by sbrannen** [comment]
Nice, @nicolaiparlog! 👍 

I'm guessing you didn't know about the [Rerunner Jupiter extension](https://github.com/artsok/rerunner-jupiter/)? 🤔

**Comment by nicolaiparlog** [comment]
Nope. :laughing: Most extensions I write come from excursions into the extension model, not from actual use cases.

Looks like @arcuri82 has at least one, maybe soon two options to achieve his goal of rerunning flaky tests in JUnit 5 - no need for native support. :wink:

**Comment by sormuras** [comment]
Related to "flaky tests": https://twitter.com/michaelbolton/status/1032125004480237568

**Comment by smoyer64** [comment]
He should sooooo change his name - https://www.youtube.com/watch?v=ADgS_vMGgzY

**Comment by anonygoose** [comment]
Hey Marc,

`import io.github.artsok.RepeatedIfExceptionsTest;

public class RerunnerBugTest {

    @RepeatedIfExceptionsTest(repeats = 3)
    void test1() {
        System.out.println("Running test 1");
    }

    @RepeatedIfExceptionsTest(repeats = 3)
    void test2() throws Exception {
        System.out.println("Running test 2");
        throw new Exception("How does this not error?  It never runs!");
    }

    @RepeatedIfExceptionsTest(repeats = 3)
    void test3() {
        System.out.println("Running test 3");
    }

}`

The above set of tests will run test1, and then skip test2 and test3.

Uploaded the full example here: https://github.com/anonygoose/rerunner-bug-example

**Comment by artsok** [comment]
Hi guys :)

I can make Pull Request with ReRun flasky mechanism, but i need information :)

**Comment by anonygoose** [comment]
Getting official support for a retry feature would be great.

> Ideally, I'd like to see something like a @RetryableTest annotation that allows one to specify: 
> * maximum number of retries
> * configurable exceptions that initiate a retry, defaulting to {Throwable.class} 
> * potentially a minimum success rate

Each of these points are configurable on the current Rerunner extension, and are useful.  Carrying them over to any official support seems sensible.

Further consideration may need giving to:
* Making retries work across the entire life-cycle, wherever the exception occurs (BeforeEach, AfterEach, etc)
* What the proper outcome of the test is.  Current rerun implementations end up with a lot of tests being reported as skipped.  This is a bit messy when trying to interpret results.

**Comment by nicolaiparlog** [comment]
It would be nice to know whether the JUnit 5 team is considering implementing this.

* If so, I will not continue working on my experiments.
* If not, this issue can be closed and @artsok and I can decide how to move forward with his extension and my snippets.

@artsok: Would you consider integrating your extension into [JUnit Pioneer](https://junit-pioneer.org)? You would get more visibility and more hands to fix issues or implement features.

**Comment by sormuras** [comment]
...now on the list for tomorrow's team call.

**Comment by benken-parasoft** [comment]
> dangerously buggy, and inactive

I wanted to second this.  I came to a similar conclusion a few weeks ago.  I had to give up on other attempts, not feeling they were entirely safe, clean, light-weight, or stable.  Instead, I ended up making my copy of RepeatedTestExtension then made it use a custom TestTemplateInvocationContext Iterator with a TestExecutionExceptionHandler that conditionally throws a TestAbortedException.  This seems to be the general approach I've seen floating around but at least I have my own light-weight solution that I have control over.  I wish astsok and pioneer the best of luck but it would definitely be nice to have a clean, official, stable, and trusted solution.

**Comment by filiphr** [comment]
Just to add my 2 cents to this issue. We also have such flaky tests and were using the Surefire `rerunFailingTestsCount` in order to circumvent this. However, since our migration to JUnit Jupiter we noticed that this does not work anymore 😄. It would be great if there is something official from JUnit or maybe the appropriate plugins (Maven, Gradle) that would still have this support.

If you ask me I think that it is a bit better if we can do this expressively in the test itself and not have it global for the entire build.

**Comment by marcphilipp** [comment]
**Team Decision:** As a first step, we think this should be maintained externally as an extension based on `@TestTemplate`. We might potentially introduce support for rerunning flaky tests in core, but then it should support all kinds of testable methods, including `@ParameterizedTests`. We'd be happy to provide guidance and review such an extension, whether in the rerunner project, junit-pioneer or another project.

**Comment by nicolaiparlog** [comment]
I've just pushed [my proof of concept](https://github.com/junit-pioneer/junit-pioneer/pull/110) to Pioneer, I'd like to implement that feature there and am interested in ideas, feedback, and even code if someone wants to write it. :smile_cat: 

@marcphilipp This can be closed, right?

**Comment by tictac-freshmint** [comment]
Surefire can do this, but unfortunately only for JUnit 4: https://maven.apache.org/surefire/maven-surefire-plugin/examples/rerun-failing-tests.html I am not sure what the equivalent for Gradle is and if one of its plugins support retries.

What we need is that in the Surefire test result XML files the information about retries are contained. We use Allure for reporting, and I believe it processes these information in order to display how often a test has been repeated and what failures occured in each retry. Hopefully Surefile will soon implement this for JUnit 5, but I did not find any issue about that. I think I will create one for that.

A solution just for within an IDE will not help us at all. Think on CI/CD and the test result reporting.

I dislike the name ¨RepeatIfFailedTest¨, four words in a long camel case. Why not just @Retry?

**Comment by ekmaughan** [comment]
My team and I are very anxious to upgrade to JUnit 5, but this has been the hang up for us. May I ask if there are any updates for this? I have noticed that the Rerunner project and Pioneer have yet to implement the ability of stacking a Retry annotation on top of @ParameterizedTests which is key to what we are wanting to accomplish.

**Comment by marcphilipp** [comment]
@ekmaughan Thanks for letting us know that this feature is important for you. I'm afraid I don't have any updates for you at this point. I agree that this would be a useful feature for integration tests that depend on external resources.

@arcuri82 Have you submitted an issue to the Surefire project for this?

**Comment by arcuri82** [comment]
@marcphilipp Someone else already did: [https://issues.apache.org/jira/browse/SUREFIRE-1584](https://issues.apache.org/jira/browse/SUREFIRE-1584)

**Comment by dlanaghen** [comment]
Hey @marcphilipp 

One more vote but we would prefer configuration like the surefire rerunFailingTestsCount feature.

We have been using JUnit5 since its experimental days with great success.  We are using it for integration testing of our API and like others, between network issues and platform issues, there is some level of unpredictability which is magnified because we are hitting our api so hard now that we can run things in parallel.

We have many tests and no one test is likely to be more "flaky" than another.  The surefire rerunFailingTestsCount was actually the ideal solution for us rather than a special annotation per test.  Ideally we would like to configure this kind of a feature for all tests.

I can see the use case for both a test specific annotation, or a configuration for all tests.

**Comment by marcphilipp** [comment]
@dlanaghen Makes sense! We could add a global config parameter for the default.

**Comment by artsok** [comment]
Hey @ekmaughan 

You can try to use my implementation with support @ParameterizedTests which I did today. As all main parameterized classes have default visibility I need to set mine owns. For that reason, I introduce a new annotation @ParameterizedRepeatedIfExceptionsTest.

Example:

```
    private ThreadLocalRandom random = ThreadLocalRandom.current();

    /**
     * By default total repeats = 1 and minimum success = 1.
     * If the test failed by this way start to repeat it by one time with one minimum success.
     *
     * This example with display name, repeated display name, 10 repeats and 2 minimum success with exceptions.
     * Exception depends on random number generation.
     */
    @DisplayName("User payment")
    @ParameterizedRepeatedIfExceptionsTest(name = "payment amount was {0}",
            repeatedName = " (Repeat {currentRepetition} of {totalRepetitions})",
            repeats = 10, exceptions = RuntimeException.class, minSuccess = 2)
    @ValueSource(ints = {40, 55, 61, 700})
    void errorParameterizedTestWithDisplayNameAndRepeatedName(int argument) {
        if (random.nextInt() % 2 == 0) {
            throw new RuntimeException("Exception in Test " + argument);
        }
    }
```

IDEA Report:
<img width="506" alt="изображение" src="https://user-images.githubusercontent.com/8410018/57184833-5630d200-6eca-11e9-8a12-61fc6b1680b3.png">
Explaining: First test with argument '40' failed. Repeat this test with first argument. Wait the sitaution when will have 2 minimum success runs. Then go to the next test with second argument and run it.  With second argument way we can see that it was no 2 minimum success follow each other. Try to repeat 10 times but wihout result, go next test with argument '61' and so on.

Allure Report (another test run):
<img width="897" alt="изображение" src="https://user-images.githubusercontent.com/8410018/57184839-6b0d6580-6eca-11e9-8fb0-caf022da7fd1.png">



Checked with this Junit version
```
<junit.jupiter.version>5.4.2</junit.jupiter.version>
<junit.platform.version>1.4.2</junit.platform.version>
```
You can try this extension:
```
<dependency>
    <groupId>io.github.artsok</groupId>
    <artifactId>rerunner-jupiter</artifactId>
    <version>2.0.1</version>
    <scope>test</scope>
</dependency>
```

The main logic you can find in ParameterizedRepeatedTestExtension.class. Source code you can find [here](https://github.com/artsok/rerunner-jupiter). Don't forget to send [star](https://github.com/artsok/rerunner-jupiter/stargazers) if u want.

**Comment by cypris** [comment]
Thanks @artsok. Looking forward to when you’ve made sure it works parallelized too!

**Comment by toby1984** [comment]
Hi,
Any progress on this ? We rely on the 'rerunFailingTestsCount' feature of maven-surefire-plugin and I just learned the hard way that it does not work with JUnit5.

**Comment by marcphilipp** [comment]
@toby1984 While we'd still like to implement this, we haven't had time to look into it more closely, yet.

One open question is if/how we should report repetitions with "allowed failures". @artsok seems to have reported them as failed, I could also envision reporting them as aborted. Or we could not report them at all and only fail the test in the end if it didn't pass after the configured number of retries.

Another open question is the lifecycle of a repetition. Spock allows to configure which parts of the setup get executed for a retry. I think we can make it simple to start with and repeat the whole lifecycle like we do for repeated and parameterized tests.

Thoughts?

**Comment by itrifonov** [comment]
VSTS has option in filters "passed on rerun", i don't know how it looks in XML but not Aborted for sure as it's a separate outcome.
![image](https://user-images.githubusercontent.com/5763935/64709387-7135d300-d4b6-11e9-9a38-320e6debcce8.png)

**Comment by marwatk** [comment]
I have a need for this with some Selenium tests that are the result of a `TestFactory` and I've been playing around with the new `InvocationInterceptor` to try to make it work:

```
    @Override
    public void interceptTestMethod( Invocation<Void> invocation,
            ReflectiveInvocationContext<Method> invocationContext,
            ExtensionContext extensionContext ) throws Throwable {
        runTest( invocation, invocationContext, extensionContext );
    }

    @Override
    public void interceptDynamicTest( Invocation<Void> invocation,
            ExtensionContext extensionContext ) throws Throwable {
        runTest( invocation, null, extensionContext );
    }

    public Void runTest( Invocation<Void> invocation,
            ReflectiveInvocationContext<Method> invocationContext,
            ExtensionContext extensionContext ) throws Throwable {

        int runCount = extensionContext.getTestMethod().isPresent()
                && extensionContext.getTestMethod().get().isAnnotationPresent( Retry.class )
                ? 3 : 1;
        
        Throwable lastError = null;
        for ( int i = 0; i < runCount; i++ ) {
            try {
                log.debug( "Running run {}/{} of [{}]",
                        i + 1, runCount, extensionContext.getDisplayName() );
                return invocation.proceed();
            }
            catch ( Throwable t ) {
                log.info( "Test run {}/{} of [{}] resulted in error {}",
                        i + 1, runCount, extensionContext.getDisplayName(), t );
                lastError = t;
            }
        }
        throw lastError;
    }
```
Which seems like a simple place to retry that doesn't report extra tests and avoids the issue of when to run them.

The problem is [`proceed` is validated](https://github.com/junit-team/junit5/blob/master/junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/execution/InvocationInterceptorChain.java#L123) to ensure it's only called once. The validation makes sense in most cases, but in this case I specifically want to run the test more than once. (When the annotation is present).

Is overriding the validation something that could be considered? Perhaps a way for the interceptor to indicate that it is purposefully re-running `proceed`? I can submit a PR if so.

**Comment by jaksa76** [comment]
I've written a library called [unreliable](https://github.com/jaksa76/unreliable) that could help here. It is not junit specific, but it does allow you to run a piece of code several times:
```java
// this will try to invoke the service 5 times
Unreliable.tenaciously(() -> unreliableService.doSomething(), 5); 
```
There are variants of this syntax that allow you to specify number of retries, exceptions to filter etc.

**Comment by marcphilipp** [comment]
@marwatk Implementing this using `InvocationInterceptor` is very limited since it does not provide the option to repeat the previous lifecycle steps. Nevertheless, I can imagine lifting the requirement to call proceed exactly once. There has also been a request to allow skipping it entirely in #2083.

**Comment by sbrannen** [comment]
@jaksa76, your library looks quite promising. Thanks for sharing.

**Comment by Seijan** [comment]
There was a step in the right direction from maven-surefire as they reenabled rerunning tests with surefire for JUnit5. This is still in a snapshot version, but maybe some of you find it helpful. https://github.com/apache/maven-surefire/pull/245
As I pointed out there, parameterized tests aren't working at all with this feature, but if you don't use this feature you should be good to go.

**Comment by Tibor17** [comment]
@Seijan 
Summary in current activities:
We implemented Surefire rerun for JUnit5 in https://issues.apache.org/jira/browse/SUREFIRE-1701 and there are ongoing issues to fix https://github.com/apache/maven-surefire/pull/245#issuecomment-548949938
The feature `skipAfterFailureCount` is reported in Surefire https://issues.apache.org/jira/browse/SUREFIRE-1710

**Comment by Tibor17** [comment]
@Seijan 
Now the re-run + @Parameterized test works. The display name in this annotation also works and you can have it in the XML report if you enable that in the configuration of the extension, see the documentation. https://issues.apache.org/jira/browse/SUREFIRE-1716
Feel free to subscribe in the mailing list and participate in the release vote:
https://lists.apache.org/thread.html/6b9987be2fbd4d82daf8aaabbd1d58dfa2a533598770e0423d0b5ee0@%3Cdev.maven.apache.org%3E

**Comment by nishantvas** [comment]
One add-on request here if the Junit5 team goes ahead with this.
It'll be nice to have a facet for this Extension as an interface as well rather than just only as an annotation.
Something like `ExecutionCondition` where a config object could be returned on basis of custom logic, say Exception type.

`@Override
  public RetryCondition evaluateExecutionCondition(ExtensionContext context, RetryCondition condition) {`

Obviously for someone who needs it to be test level Annotation will then provide values to have the config object via annotation fields

This way, if someone intends to have global support for their E2E tests (I know this sounds ghastly) but,
> We have many tests and no one test is likely to be more "flaky" than another. 

The Extension can be registered using autodetect config and have it tailored for Exceptions coming from all tests

**Comment by marcphilipp** [comment]
FWIW Gradle now also supports this via the [Test Retry Gradle Plugin](https://blog.gradle.org/gradle-flaky-test-retry-plugin).

**Comment by JohnArrowwood** [comment]
+1 on the need for re-running flaky end-to-end tests.  My personal example:  Change a user's password in AD.  Then confirm that the new password works.  But since the change was written directly to the primary node but the bind happens by connecting to the load balancer, there is a good chance that the attempt to bind will hit the secondary node before the password change was able to propagate to it, and when that happens, it fails.

Adding a fixed-length timeout to try and give the data time to propagate is lame.  Adding a replication check to explicitly wait for the data to propagate is complicated.  And adding a retry loop to every place where things might go wrong is . . , yeah, right, not gonna happen.  But adding a config option that says "If it fails, retry, up to N times", that would really make my life MUCH better!

**Comment by nipafx** [comment]
Here are a few options to rerun failing tests:

* [Unreliable](https://github.com/jaksa76/unreliable) by @jaksa76 
* [rerunner-jupiter](https://github.com/artsok/rerunner-jupiter) by @artsok 
* [`@RetryingTest](https://junit-pioneer.org/docs/retrying-test/) in JUnit Pioneer
    * Disclosure A: I'm one of the maintainers of Pioneer
    * Disclosure B: We really like those other projects and are in the process of ~~stealing~~ being inspired by their ideas

**Comment by sbrannen** [comment]
> [Unreliable](https://github.com/jaksa76/unreliable) by @jaksa76

You may also be interested in the `RetryTemplate` from [Spring Retry](https://github.com/spring-projects/spring-retry#imperative-example) if you are looking for a programmatic approach similar to `UnreliableService`.

**Comment by asafm** [comment]
> Here are a few options to rerun failing tests:
> 
>     * [Unreliable](https://github.com/jaksa76/unreliable) by @jaksa76
> 
>     * [rerunner-jupiter](https://github.com/artsok/rerunner-jupiter) by @artsok
> 
>     * [`@RetryingTest`](https://junit-pioneer.org/docs/retrying-test/) in JUnit Pioneer
>       
>       * Disclosure A: I'm one of the maintainers of Pioneer
>       * Disclosure B: We really like those other projects and are in the process of ~stealing~ being inspired by their ideas

From I see, none of them support applying it to all, which uses the standard `@Test` annotation, right?

**Comment by marcphilipp** [comment]
Which build tool are you using? Both Maven and Gradle support retrying on the build level.

**Comment by asafm** [comment]
> Which build tool are you using? Both Maven and Gradle support retrying on the build level.

I'm referring specifically using JUnit with the tools list above and not applying it in build tool level.

**Comment by marcphilipp** [comment]
> From I see, none of them support applying it to all, which uses the standard `@Test` annotation, right?

That's correct.

**Comment by nipafx** [comment]
@asafm Unless I'm missing something, there's no way to rerun failed tests from within Jupiter without providing a replacement for `@Test`. If I'm right, you preferred solution doesn't exist.

**Comment by deepakab03** [comment]
Hope this is added to the main JUnit Jupiter soon.
Another use-case for this is when there are unit tests that depend on timeout conditions and said tests run on CI nodes and the nodes are busy or under-powered. These tests randomly fail on assertion conditions like a particular method should have been called twice, but on CI node it is called thrice because the calling thread went to sleep for a bit, messing up the timeout condition.. if this were to be repeated a couple of times, the chances of it passing would be much higher, increasing the calls to thrice is not the correct option but we are forced to do this when the tests runs on the CI node by means of some conditions which is ugly..

**Comment by asafm** [comment]
@nipafx You are correct, solution doesn't not exists. I want Jupiter to support the ability to do so :)

**Comment by marcphilipp** [comment]
It would be relatively easy to implement this in the platform. However, it might not satisfy the needs of all users since static state, threads that leaked from prior tests and might have caused the flakiness doesn't get reset.

**Comment by hakanai** [comment]
Another vote here.

I looked at Gradle's flaky re-runner and I'm not sure that having the setting be project-global is very great, especially when the vast majority of the tests are expected to be rock-solid.

rerunner-jupiter's and JUnit Pioneer's annotations didn't really do it for us either - they both introduced a new annotation for flaky tests which _replaces_ the `@Test` annotation. This is less than great for two separate reasons:
(1) When we want to _disable_ a test, we don't go and replace the `@Test` annotation with a `@DisabledTest` annotation - we _add_ the `@Disabled` annotation next to the `@Test` annotation. Having flakiness annotations behave differently is inconsistent.
(2) When you want to mark a `@ParameterizedTest` as flaky, now you need _yet another_ annotation. (And this will compound the more things you want to annotate about the test, or the more ways you have of running a test.)

So we feel that an annotation to mark a test as flaky should sit in a separate annotation and set out to implement that.

I was considering creating a new ticket for the extension point I'd want to implement this (something like an `AroundEachCallback`), but figured maybe it was probably more productive to just vote on the feature itself because it would save a load of hassle if it were just a feature.

**Comment by marcphilipp** [comment]
> I looked at Gradle's flaky re-runner and I'm not sure that having the setting be project-global is very great, especially when the vast majority of the tests are expected to be rock-solid.

Did you see that you can specify custom annotations or class name pattern for those tests that should be retried?

**Comment by hakanai** [comment]
> > I looked at Gradle's flaky re-runner and I'm not sure that having the setting be project-global is very great, especially when the vast majority of the tests are expected to be rock-solid.
> 
> Did you see that you can specify custom annotations or class name pattern for those tests that should be retried?

I did not, no. Does it also let you have a different number of retries per test?

**Comment by marcphilipp** [comment]
> I did not, no. Does it also let you have a different number of retries per test?

No, the number of retries can only be configured globally.

> If only an `InvocationInterceptor` could just call `proceed()` twice, this would be much, much easier.

I stand by what I said previously in https://github.com/junit-team/junit5/issues/1558#issuecomment-546594703.

**Comment by SandersAaronD** [comment]
> It would be relatively easy to implement this in the platform. However, it might not satisfy the needs of all users since static state, threads that leaked from prior tests and might have caused the flakiness doesn't get reset.

I'd like to add +1 for doing this.

To add to the use cases: We have a bunch of E2E tests. Some are working, some are flaky, and some are broken. In any given test run, it's difficult to tell which are flaky and which are actually broken, so I have to rerun them manually to sort those out. Right now I will run my entire test suite 3-5 times and manually look at the reports side by side to see which tests fail uniformly, and which fail only sometimes.

A broken test indicates, probably, a logic error somewhere. A flaky test indicates a latency, availability or concurrency issue. They are very different things, and I need to go looking for the root issue in different places, and being able to see the difference matters.

We already have to manage static state and threading issues when doing repeated runs of a test or using test concurrency in junit, it's not clear to me that facing those issues is a good reason not to put an opt-in feature in since those are already common for use cases already in the package.

**Comment by SandersAaronD** [comment]
> > I did not, no. Does it also let you have a different number of retries per test?
> 
> No, the number of retries can only be configured globally.
> 
> > If only an `InvocationInterceptor` could just call `proceed()` twice, this would be much, much easier.
> 
> I stand by what I said previously in [#1558 (comment)](https://github.com/junit-team/junit5/issues/1558#issuecomment-546594703).

You can specify a custom task that only runs tests for certain tags and set the retry block in the gradle plugin there. I have retries set of my regression tests but not smoke tests, for example, since they are triggered from different tasks.

**Comment by Michael1993** [comment]
Technically, I think we (JUnit Pioneer) could support that. You would be able to write something like this:
```java
@Test
@Flaky(maxRetries = 3)
void test() {
  // some flaky test code here
}
```

It would create results like this:
![image](https://user-images.githubusercontent.com/4085611/151206998-bd7a191d-62fb-4335-994e-4d21d3b32b6c.png)

Personally, I think that looks a bit ugly... but if this is good enough and there is demand for it, we will host it (I'll ask the other maintainers but I'm fairly confident they'll agree).

**Comment by ksobolew** [comment]
Yeah, this is inconsistent with the general pattern. More "regular" way would be to add `@FlakyTest`, `@FlakyParameterizedTest` etc., which may be even worse.

**Comment by wyhasany** [comment]
It passed more then 3 years and there is no reliable solution to retry `@ParameterizedTest` or `@TestFactory`. There are only ones working with build tools (surefire for maven and gradle retry plugin). Maybe console-launcher could be configurable to rerun failing tests?

**Comment by nipafx** [comment]
@iNikem Yes, that is the correct. I explain the reason [in a related issue](https://github.com/junit-pioneer/junit-pioneer/issues/405#issuecomment-1306849451).

**Comment by Avinm** [comment]
> It passed more then 3 years and there is no reliable solution to retry @ParameterizedTest or @TestFactory.

I am here to echo this on behalf of several of my team members. Would there be anything we could consider to inch closer to resolving this (and essentially https://github.com/junit-pioneer/junit-pioneer/issues/586).

I could try contributing towards an implementation if junit owners can align on and provide a direction.
FWIW, I think testng has `IRetryAnalyzer` which helps to neatly solve this problem.

Ref: https://testng.org/doc/documentation-main.html#rerunning

**Comment by benken-parasoft** [comment]
> I ended up making my own copy of RepeatedTestExtension

JUnit 5.10.0 now has "Failure threshold for @RepeatedTest."  No more need for my hacked copy of RepeatedTestExtension.  Thanks!

**Comment by JonathanBerkeley** [comment]
If there was successThreshold for @RepeatedTest this would be solved!

**Comment by carusology** [comment]
An additional use case has come up when writing LLM-based AI integration tests.

When an LLM-based AI is asked to return a response with well-formatted data, it can do so reliably the vast majority of the time. However, it isn't truly deterministic - even if provided the exact same user and system same prompts. Having first class support for retrying would way to accommodate this unpredictability.

It would be even better if it could retry based upon specific exception types. For example, it could only retry parsing errors because the LLM didn't respond in the expected well-formatted data specified in the system prompts.

**Comment by pan3793** [comment]
I'm wondering how to achieve the scalatest's `eventually` functionality in JUnit5? This is extremely useful to tackle flaky test assertions or wait some resources to start before moving to next steps.

```
test("test case contains flaky assertion") {
  // ... some logic maybe with assertion
  
  eventually(timeout(30 seconds), interval(100 millis)) {
     // assertion that may fail
  }

  // ... some logic maybe with assertion
}
```

Ref: http://doc.scalatest.org/1.8/org/scalatest/concurrent/Eventually.html

**Comment by quiram** [comment]
Hi @pan3793,

I think for this the easiest is using a tool like [awaitility](https://github.com/awaitility/awaitility). We use it in our projects for precisely this reason and it works fairly well.

**Comment by ArvindJoshi-okta** [comment]
All of the above mentioned workaround options have annotations (for rerunning flaky test) at a method level. I am looking for an option to specify the flakiness retry for the entire code base, could be at class level on a base class. Any suggestions for that? Thanks!

**Comment by Avinm** [comment]
@ArvindJoshi-okta:
I was able to achieve it (in an admittedly hacky manner) using an InvocationInterceptor:
````
public void interceptTestTemplateMethod(Invocation<Void> invocation,
                                            ReflectiveInvocationContext<Method> invocationContext,
                                            ExtensionContext extensionContext) throws Throwable {
    try {
            invocation.proceed();
    } catch (Throwable initialEx) {
       if(shouldRetry(initialEx) {
               invocationContext.getExecutable().invoke(
                        invocationContext.getTarget().get(), invocationContext.getArguments().toArray()
              ); // Invokes your method again, consider looping with try catch if multiple retries are needed
       }
    }
}

```

One downfall of this approach is that it does not execute the @Before* methods again though

**Comment by benken-parasoft** [comment]
> Adding successThreshold to @RepeatedTest would solve this.

I just wanted to highlight this again.  It seems reasonable.

**Comment by nipafx** [comment]
> Adding successThreshold to @RepeatedTest would solve this.

I want to offer a counterpoint to this sentiment. At the moment `@RepeatedTest`s only succeed if all executions succeed, which makes the annotations fundamentally unfit to carry flaky tests across the finishing line. Adding `successThreshold` changes that just as fundamentally and opens the door to an entirely new use case. A use case that, as we have discovered in JUnit Pioneer, quickly demands follow-up additions:

* fixed pause times
* backoff strategies
* differentiate by exception type
* probably more I don't remember right now

There's nothing wrong with these features but IMO it doesn't make sense to take the first step down that road if one is not willing to travel it further.

So, as I see it, the decision is not "can we add `successThreshold` to `@RepeatedTest`?" but "do we want to turn `@RepeatedTest` into an annotation that (properly) supports rerunning flaky tests?".
